baseapp/internal/protocompat/protocompat.go

```diff
@@ -0,0 +1,202 @@
+package protocompat
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+
+	gogoproto "github.com/cosmos/gogoproto/proto"
+	"google.golang.org/grpc"
+	proto2 "google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/reflect/protoreflect"
+	"google.golang.org/protobuf/reflect/protoregistry"
+	"google.golang.org/protobuf/runtime/protoiface"
+
+	"github.com/cosmos/cosmos-sdk/codec"
+)
+
+var (
+	gogoType    = reflect.TypeOf((*gogoproto.Message)(nil)).Elem()
+	protov2Type = reflect.TypeOf((*proto2.Message)(nil)).Elem()
+)
+
+type Handler = func(ctx context.Context, message protoiface.MessageV1) (protoiface.MessageV1, error)
```

Comment:
```suggestion
type Handler = func(ctx context.Context, request, response protoiface.MessageV1) error
```

We discussed changing to this signature in our last call. Strongly recommend this because it matches the signature of `ClientConnInterface.Invoke`. The signature that this PR has can be a wrapper that looks up the response type, but generally there should be no need to look up the response type because `Invoke` will provide it

Reply:
If we follow this API we'd need to merge on the resp also